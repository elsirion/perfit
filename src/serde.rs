use std::fmt;

use serde::{de, Deserializer};

pub fn deserialize_opt_f64_from_empty_string<'de, D>(
    deserializer: D,
) -> Result<Option<f64>, D::Error>
where
    D: Deserializer<'de>,
{
    struct OptF64Visitor;

    impl<'de> de::Visitor<'de> for OptF64Visitor {
        type Value = Option<f64>;

        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
            formatter.write_str("null, an empty string, or a float")
        }

        fn visit_some<D>(self, deserializer: D) -> Result<Option<f64>, D::Error>
        where
            D: Deserializer<'de>,
        {
            deserializer.deserialize_str(self)
        }

        // For directly handling nulls, leading to None
        fn visit_none<E>(self) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Ok(None)
        }
        // Handle missing value as None
        fn visit_unit<E>(self) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Ok(None)
        }

        // Handle missing value as None
        fn visit_f64<E>(self, value: f64) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Ok(Some(value))
        }

        // Handle empty string as None
        fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            if value.is_empty() {
                Ok(None)
            } else {
                match value.parse::<f64>() {
                    Ok(v) => Ok(Some(v)),
                    Err(_) => Err(E::custom("Expected a float")),
                }
            }
        }
    }

    deserializer.deserialize_option(OptF64Visitor)
}

pub mod custom_rfc3339_option {
    use serde::Serializer;
    use time::OffsetDateTime;

    #[allow(clippy::wildcard_imports)]
    use super::*;

    pub fn serialize<S: Serializer>(
        option: &Option<OffsetDateTime>,
        serializer: S,
    ) -> Result<S::Ok, S::Error> {
        time::serde::rfc3339::option::serialize(option, serializer)
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<time::OffsetDateTime>, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct Visitor;

        impl<'de> de::Visitor<'de> for Visitor {
            type Value = Option<OffsetDateTime>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("null, an empty string, or a float")
            }

            fn visit_some<D>(self, deserializer: D) -> Result<Option<OffsetDateTime>, D::Error>
            where
                D: Deserializer<'de>,
            {
                deserializer.deserialize_str(self)
            }

            // For directly handling nulls, leading to None
            fn visit_none<E>(self) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                Ok(None)
            }
            // Handle missing value as None
            fn visit_unit<E>(self) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                Ok(None)
            }

            // Handle empty string as None
            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                if value.is_empty() {
                    Ok(None)
                } else {
                    Ok(Some(
                        OffsetDateTime::parse(
                            value,
                            &time::format_description::well_known::Rfc3339,
                        )
                        .map_err(|e| E::custom(e.to_string()))?,
                    ))
                }
            }
        }

        deserializer.deserialize_option(Visitor)
    }
}
