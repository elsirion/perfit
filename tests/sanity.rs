mod common;

use color_eyre::Result;
use insta_cmd::get_cargo_bin;
use serde::Deserialize;
use tracing::info;

use crate::common::PerfitdFixture;

trait ExpressionExt {
    fn read_json<T>(self) -> Result<T>
    where
        T: for<'de> Deserialize<'de>;

    fn read_json_value(self) -> Result<serde_json::Value>
    where
        Self: Sized,
    {
        self.read_json()
    }
}

impl ExpressionExt for duct::Expression {
    fn read_json<T>(self) -> Result<T>
    where
        T: for<'de> Deserialize<'de>,
    {
        let s = self.stdout_capture().read()?;
        let v: T = serde_json::from_str(&s)?;
        Ok(v)
    }
}

#[derive(Deserialize)]
struct NewAccountOutput {
    #[allow(unused)]
    account_id: String,
    access_token: String,
}

#[tokio::test(flavor = "multi_thread")]
async fn sanity_post_data_points() -> Result<()> {
    common::init_logging()?;

    let fixture = PerfitdFixture::new().await?;

    let addr = fixture.addr()?;

    let root_access_token = fixture.root_access_token_str();

    fixture
        .run(async {
            info!("Staring test");
            let bin = get_cargo_bin("perfit");
            tokio::task::spawn_blocking(move || -> Result<_> {
                let NewAccountOutput {
                    account_id: _,
                    access_token,
                } = duct::cmd!(&bin, "account", "new")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", root_access_token)
                    .read_json()?;

                let metric_id: String = duct::cmd!(&bin, "metric", "new")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", &access_token)
                    .read_json()?;

                insta::assert_yaml_snapshot!("no data points", duct::cmd!(&bin, "metric", "get")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", &access_token)
                    .env("PERFIT_METRIC", &metric_id)
                    .read_json_value()?, {
                    "[].t" => "[ts]",
                });
                duct::cmd!(&bin, "post", "11")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", &access_token)
                    .env("PERFIT_METRIC", &metric_id)
                    .run()?;

                insta::assert_yaml_snapshot!("one data point", duct::cmd!(&bin, "metric", "get")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", &access_token)
                    .env("PERFIT_METRIC", &metric_id)
                    .read_json_value()?, {
                    "[].t" => "[ts]",
                });

                duct::cmd!(&bin, "run", "sleep", ".01")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", &access_token)
                    .env("PERFIT_METRIC", &metric_id)
                    .run()?;

                duct::cmd!(&bin, "post", "12")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", &access_token)
                    .env("PERFIT_METRIC", &metric_id)
                    .run()?;

                insta::assert_yaml_snapshot!("three data points", duct::cmd!(&bin, "metric", "get")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", &access_token)
                    .env("PERFIT_METRIC", &metric_id)
                    .read_json_value()?, {
                        "[].t" => "[ts]",
                        "[1].v" => "[value]",
                    }
                );

                Ok(())
            })
            .await??;
            Ok(())
        })
        .await
}

#[tokio::test(flavor = "multi_thread")]
async fn sanity_user_mgmt() -> Result<()> {
    common::init_logging()?;

    let fixture = PerfitdFixture::new().await?;

    let addr = fixture.addr()?;

    let root_access_token = fixture.root_access_token_str();

    fixture
        .run(async {
            info!("Staring test");
            let bin = get_cargo_bin("perfit");
            tokio::task::spawn_blocking(move || -> Result<_> {
                let NewAccountOutput {
                    account_id: _,
                    access_token,
                } = duct::cmd!(&bin, "account", "new")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", root_access_token)
                    .read_json()?;

                let NewAccountOutput {
                    account_id: _,
                    access_token: _,
                } = duct::cmd!(&bin, "token", "new")
                    .env("PERFIT_SERVER", format!("http://{}", addr))
                    .env("PERFIT_ACCESS_TOKEN", access_token)
                    .read_json()?;

                Ok(())
            })
            .await??;
            Ok(())
        })
        .await
}
