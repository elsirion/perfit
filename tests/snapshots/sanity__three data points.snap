---
source: tests/sanity.rs
expression: "duct::cmd!(&bin, \"metric\",\n                        \"get\").env(\"PERFIT_SERVER\",\n                    format!(\"http://{}\",\n                        addr)).env(\"PERFIT_ACCESS_TOKEN\",\n                &access_token).env(\"PERFIT_METRIC\",\n            &metric_id).read_json_value()?"
---
- t: "[ts]"
  v: 11
- t: "[ts]"
  v: "[value]"
- t: "[ts]"
  v: 12
